src/aks-preview/azext_aks_preview/_podidentity.py

```diff
@@ -0,0 +1,151 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+import time
+
+from knack.log import get_logger
+from knack.util import CLIError
+
+from ._client_factory import get_auth_management_client
+from ._consts import (CONST_MANAGED_IDENTITY_OPERATOR_ROLE,
+                      CONST_MANAGED_IDENTITY_OPERATOR_ROLE_ID)
+from ._roleassignments import add_role_assignment
+
+logger = get_logger(__name__)
+
+
+def _is_pod_identity_addon_enabled(instance):
+    if not instance:
+        return False
+    if not instance.pod_identity_profile:
+        return False
+    return bool(instance.pod_identity_profile.enabled)
+
+
+def _ensure_pod_identity_addon_is_enabled(instance):
+    if not _is_pod_identity_addon_enabled(instance):
+        raise CLIError('The pod identity addon is not enabled for this managed cluster yet.\n'
+                       'To enable, run "az aks update --enable-pod-identity')
+
+
+def _ensure_pod_identity_kubenet_consent(network_profile, pod_identity_profile, customer_consent):
+    if not network_profile or not network_profile.network_plugin:
+        # invalid data
+        return
+    if network_profile.network_plugin.lower() != 'kubenet':
+        # not kubenet, no need to check
+        return
+
+    if customer_consent is None:
+        # no set this time, read from previous value
+        customer_consent = bool(
+            pod_identity_profile.allow_network_plugin_kubenet)
+
+    if not customer_consent:
+        raise CLIError(
+            "--enable-pod-identity-with-kubenet is required for "
+            "enabling pod identity addon when using Kubenet network plugin")
+    pod_identity_profile.allow_network_plugin_kubenet = True
+
+
+def _fill_defaults_for_pod_identity_exceptions(pod_identity_exceptions):
+    if not pod_identity_exceptions:
+        return
+
+    for exc in pod_identity_exceptions:
+        if exc.pod_labels is None:
+            # in previous version, we accidentally allowed user to specify empty pod labels,
+            # which will be converted to `None` in response. This behavior will break the extension
+            # when using 2021-10-01 version. As a workaround, we always back fill the empty dict value
+            # before sending to the server side.
+            exc.pod_labels = dict()
+
+
+def _fill_defaults_for_pod_identity_profile(pod_identity_profile):
+    if not pod_identity_profile:
+        return
+
+    _fill_defaults_for_pod_identity_exceptions(pod_identity_profile.user_assigned_identity_exceptions)
+
+
+def _update_addon_pod_identity(
+    instance,
+    enable,
+    pod_identities=None,
+    pod_identity_exceptions=None,
+    allow_kubenet_consent=None,
+    models=None,
```

Comment:
nit: we should properly annotate the type for the `models` parameter here

Reply:
Sure.

Reply:
Updated in commit [#9e1827e](https://github.com/Azure/azure-cli-extensions/pull/4243/commits/9e1827e8f5150a8f3e3884c433c9d9e35f668c3d).